internal/net/udp/packet_conn.go

```diff
@@ -0,0 +1,399 @@
+// SPDX-FileCopyrightText: 2023 The Pion community <https://pion.ly>
+// SPDX-License-Identifier: MIT
+
+// Package udp implements DTLS specific UDP networking primitives.
+// NOTE: this package is an adaption of pion/transport/udp. If possible, the
+// updates made in this repository will be reflected back upstream. If not, it
+// is likely that this will be moved to a public package in this repository.
+package udp
+
+import (
+	"context"
+	"errors"
+	"net"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	idtlsnet "github.com/pion/dtls/v2/internal/net"
+	dtlsnet "github.com/pion/dtls/v2/pkg/net"
+	"github.com/pion/transport/v2/deadline"
+)
+
+const (
+	receiveMTU           = 8192
+	defaultListenBacklog = 128 // same as Linux default
+)
+
+// Typed errors
+var (
+	ErrClosedListener      = errors.New("udp: listener closed")
+	ErrListenQueueExceeded = errors.New("udp: listen queue exceeded")
+)
+
+// listener augments a connection-oriented Listener over a UDP PacketConn
+type listener struct {
+	pConn *net.UDPConn
+
+	accepting      atomic.Value // bool
+	acceptCh       chan *PacketConn
+	doneCh         chan struct{}
+	doneOnce       sync.Once
+	acceptFilter   func([]byte) bool
+	connResolver   func([]byte) (string, bool)
+	connIdentifier func([]byte) (string, bool)
+
+	connLock sync.Mutex
+	conns    map[string]*PacketConn
+	connWG   *sync.WaitGroup
```

Comment:
Thanks for catching! This was copied over, so I wanted to check and see if there was any reason why this pattern was used or if it was just an accident. `pion/transport/udp` was recently copied from `pion/udp` to its new location. It appears that some changes were made during that process.
- This PR migrated the package: https://github.com/pion/transport/pull/234/files
- This commit was made prior to the migration, but is not reflected there: https://github.com/pion/udp/commit/2589407f52c94a7e1ec63d17df744eabc4fb1ea5

That commit was made well before the migration, and appears to fix a data race issue, so I believe it was erroneously dropped. I would recommend that we pull it back in, which unfortunately also means that we will need `pion/udp/pkg/sync` (https://github.com/pion/udp/tree/master/pkg/sync), so I'll go ahead and pull that in here. I'll also open an issue to see if we can get more clarity on whether the change was intentional, but I'd prefer to stay as true to the working functionality as possible, then make improvements in smaller batches as needed.

Reply:
oh yeah ignore this one. i meant to delete comments that I realized were on existing code

Reply:
I'm actually going to opt not to pull the `pion/udp/pkg/sync` in here quite yet until I can get some more clarity on why it wasn't migrated (asked in https://github.com/pion/transport/issues/57#issuecomment-1694528604). For now, I am going to just update with your suggestion to use a `sync.WaitGroup` rather than `*sync.WaitGroup` 👍🏻